src/include/platform/OTARequestorStorage.h

```diff
@@ -0,0 +1,42 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <lib/support/Span.h>
+
+namespace chip {
+namespace DeviceLayer {
+
+class OTARequestorStorage
+{
+public:
+    using ProviderLocationType = app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type;
+
+    virtual ~OTARequestorStorage() {}
+
+    virtual CHIP_ERROR StoreProviderLocation(const ProviderLocationType & provider) = 0;
+    virtual CHIP_ERROR LoadProviderLocation(ProviderLocationType & provider)        = 0;
```

Comment:
Suggest using `NodeId`,  `FabricIndex` and `EndpointId` base types from `lib/core/DataModelTypes.h` rather than trying to use the cluster-specific  `ProviderLocationType` which causes a circular dependency between platform code (that should be cluster-data-representation-agnostic)  and clusters

Reply:
Ok, I thought I could reuse the serialization code and save some flash that way (I observe that our TLV framework isn't the lightest one). Should we encode these values with a custom TLV or is raw binary data acceptable?

Reply:
You can reuse the serialization logic in the *implementation* side (.cpp), that will not be under platform/, but rather under examples/ or some other non-core location. The key aspect is avoiding cluster-specific generated-code dependencies in public APIs

Reply:
Ok, but just FYI :D
```
$ git grep '#include <app-common' src/include/
src/include/platform/AttributeList.h:#include <app-common/zap-generated/cluster-objects.h>
src/include/platform/ConfigurationManager.h:#include <app-common/zap-generated/cluster-objects.h>
src/include/platform/ConnectivityManager.h:#include <app-common/zap-generated/cluster-objects.h>
src/include/platform/DiagnosticDataProvider.h:#include <app-common/zap-generated/cluster-objects.h>
src/include/platform/OTARequestorDriver.h:#include <app-common/zap-generated/cluster-objects.h>
src/include/platform/OTARequestorInterface.h:#include <app-common/zap-generated/cluster-objects.h>
src/include/platform/ThreadStackManager.h:#include <app-common/zap-generated/cluster-objects.h>
```
I guess people tend to ignore some engineering practices to avoid yet another level of indirection, but I get your point.

Reply:
> The key aspect is avoiding cluster-specific generated-code dependencies in public APIs

Why?  That is, why are cluster-defined things not allowed in public APIs?

Reply:
I would argue it's not simpler. If we want to enforce the rule that the code in `src` can't directly operate on Data Model structures (which is already extensively violated), then we need some sort of translation code between Data Model structures and core structures, defined in `src/`. The translation code can't be unit tested by the current framework, so I think that by allowing the core code to operate on Data Model structures (auto-generated) we actually increase the unit test coverage, right? Not to mention that we reduce the code size that way.